genlm/control/potential/built_in/canonical.py

```
@@ -0,0 +1,453 @@
+        self._decode = _decode
+        self.V = len(_decode)  # token vocabulary size
+
+        # priority dict might still be useful if merges aren't strictly ordered
+        # or for potential future optimizations, keep it for now.
+        # self.priority = {(u, v): -i for i, (u, v, _) in enumerate(self._merges)}
+        self.make_derivation_table()  # Call the rewritten method
+
+        self.__left_spine, max_left_spine_width = self._left_spine_table()
+        self.__right_spine, max_right_spine_width = self._right_spine_table()
+
+        self.left_spine_vector = self.vectorize_spine(
+            self.__left_spine, max_left_spine_width
+        )
+        self.right_spine_vector = self.vectorize_spine(
+            self.__right_spine, max_right_spine_width
+        )
+
+        self.indices = np.array(
+            [
+                (index, j)
+                for index in range(self.V)
+                for j in range(len(self.__left_spine[index]) - 1)
+            ]
+        )
+
+        self.vector_r = self.left_spine_vector[self.indices[:, 0], self.indices[:, 1]]
+        self.vector_rp = self.left_spine_vector[
+            self.indices[:, 0], self.indices[:, 1] + 1
+        ]
+
+        tmp = sp.dok_matrix((self.V, self.V), dtype=np.int32)
+        for u, v, uv in _merges:
+            tmp[u, v] = uv + 1  # +1 to avoid zero-indexing
+
+        self.parent_l_matrix = tmp.tocsr()
+        self.parent_l_matrix = self.parent_l_matrix[:, self.vector_r]
+
+        self.eos_token_ids = set(eos_token_ids)
+        self.overrides = defaultdict(lambda: set())
+
+    def __call__(self, context):
+        if context == ():
+            mask = np.ones(self.V, dtype=bool)
+        else:
+            (_, last_token) = context
+            try:
+                left_id = self._encode[last_token]  # Get the ID of the last token
+            except KeyError:
+                # If last token is unknown, disallow everything except EOS, maybe raise error?
```

I'd raise an error here